Transitive C dependencies are not linked

https://github.com/iphydf/repro170 (output: https://travis-ci.org/iphydf/repro170)

Removing the intermediate `cbits-indirect` fixes it. Adding an indirection through `cbits-indirect` such that e.g. `get_thing` is called from within `cbits-indirect` and `MyModule` only calls that indirection function fixes the linker (ld) error, but replaces it with a dynamic linker (ld.so) error.

```python
load("@io_tweag_rules_haskell//haskell:haskell.bzl", "haskell_binary", "haskell_library")

cc_library(
    name = "cbits-indirect",
    srcs = ["cbits/impl.c"],
)

cc_library(
    name = "cbits",
    deps = ["cbits-indirect"],
)

haskell_library(
    name = "mypkg",
    srcs = ["src/MyModule.hs"],
    prebuilt_dependencies = ["base"],
    src_strip_prefix = "src",
    deps = [":cbits"],
)

haskell_binary(
    name = "mytest",
    srcs = ["test/Main.hs"],
    prebuilt_dependencies = ["base"],
    src_strip_prefix = "test",
    deps = [":mypkg"],
)
```

For me your updated example from the repo fails at linking of `//mypkg:mytest`. Like so:

```
ERROR: /home/repro170/mypkg/BUILD.bazel:22:1: Linking mytest failed (Exit 1)
bazel-out/k8-fastbuild/bin/mypkg/libmypkg_Scbits_libcbits.so: undefined reference to `real_set_thing'
bazel-out/k8-fastbuild/bin/mypkg/libmypkg_Scbits_libcbits.so: undefined reference to `real_get_thing'
```

Maybe there is some difference in how the linker works, I'm not sure. The problem is that `cbits` cannot find `cbits-indirect`. And the bigger problem is that when we have a C dependency, we know nothing about its transitive dependencies. We basically have a `DefaultInfo` provider in that case, and it does not contain such info. I thought there may be a field with runfiles or something, but there is nothing useful.

`CcSkylarkApiProvider` does not contain info about transitive dependencies and it's not visible to us anyway.

I need to find out what cc rules do in such cases.

--- a/haskell/private/dependencies.bzl
+++ b/haskell/private/dependencies.bzl
@@ -21,37 +21,48 @@ def _is_shared_library(f):
     """
     return f.extension in ["so", "dylib"] or f.basename.find(".so.") != -1
 
-def _mangle_solib(ctx, label, solib, preserve_name):
-    """Create a symlink to a dynamic library, with a longer name.
+def _is_static_library(f):
+    """Check if the given File is a static library.
 
-    The built-in cc_* rules don't link against a shared library
+    Args:
+      f: The File to check.
+
+    Returns:
+      Bool: True if the given file `f` is a static library, False otherwise.
+    """
+    return f.extension in ["a"]
+
+def _mangle_lib(ctx, label, lib, preserve_name):
+    """Create a symlink to a library, with a longer name.
+
+    The built-in cc_* rules don't link against a library
     directly. They link against a symlink whose name is guaranteed to be
-    unique across the entire workspace. This disambiguates dynamic
-    libraries with the same soname. This process is called "mangling".
+    unique across the entire workspace. This disambiguates
+    libraries with the same name. This process is called "mangling".
     The built-in rules don't expose mangling functionality directly (see
     https://github.com/bazelbuild/bazel/issues/4581). But this function
     emulates the built-in dynamic library mangling.
 
     Args:
       ctx: Rule context.
-      label: the label to use as a qualifier for the dynamic library name.
-      solib: the dynamic library.
-      preserve_name: Bool, whether given `solib` should be returned unchanged.
+      label: the label to use as a qualifier for the library name.
+      solib: the library.
+      preserve_name: Bool, whether given `lib` should be returned unchanged.
 
     Returns:
-      File: the created symlink or the original solib.
+      File: the created symlink or the original lib.
     """
 
     if preserve_name:
-        return solib
+        return lib
 
     components = [c for c in [label.workspace_root, label.package, label.name] if c]
     qualifier = "/".join(components).replace("_", "_U").replace("/", "_S")
-    qualsolib = ctx.actions.declare_file("lib" + qualifier + "_" + solib.basename)
+    quallib = ctx.actions.declare_file("lib" + qualifier + "_" + lib.basename)
 
-    ln(ctx, solib, qualsolib)
+    ln(ctx, lib, quallib)
 
-    return qualsolib
+    return quallib
 
 def gather_dep_info(ctx):
     """Collapse dependencies into a single `HaskellBuildInfo`.
@@ -131,7 +142,7 @@ def gather_dep_info(ctx):
                         f:
                         # If the provider is CcSkylarkApiProviderHacked, then the .so
                         # files come from haskell_cc_import.
-                        _mangle_solib(ctx, dep.label, f, CcSkylarkApiProviderHacked in dep)
+                        _mangle_lib(ctx, dep.label, f, CcSkylarkApiProviderHacked in dep)
                         for f in dep.files.to_list()
                         if _is_shared_library(f)
                     },

--- a/haskell/private/dependencies.bzl
+++ b/haskell/private/dependencies.bzl
@@ -125,6 +125,19 @@ def gather_dep_info(ctx):
                 direct_prebuilt_deps = set.mutable_insert(acc.direct_prebuilt_deps, pkg),
             )
         else:
+            # The final link of a library must include all static
+            # libraries we depend on, including transitives ones.
+            # Theses libs are provided in `dep.cc.libs` attribute.
+            transitive_static_deps = {}
+
+            # Transitives static dependencies
+            if hasattr(dep, "cc"):
+                transitive_static_deps = {
+                    name: _mangle_lib(ctx, dep.label, name, CcSkylarkApiProviderHacked in dep)
+                    for name in dep.cc.libs.to_list()
+                    if _is_static_library(name)
+                }
+
             # If not a Haskell dependency, pass it through as-is to the
             # linking phase.
             acc = HaskellBuildInfo(
@@ -146,6 +159,7 @@ def gather_dep_info(ctx):
                         for f in dep.files.to_list()
                         if _is_shared_library(f)
                     },
+                    transitive_static_deps,
                 ),
                 direct_prebuilt_deps = acc.direct_prebuilt_deps,
             )

--- a/tests/indirect-link/BUILD.bazel
+++ b/tests/indirect-link/BUILD.bazel
@@ -1,0 +1,29 @@
+load("@io_tweag_rules_haskell//haskell:haskell.bzl", "haskell_binary", "haskell_library", "haskell_import")
+
+haskell_import(name = "base")
+
+cc_library(
+    name = "cbits-indirect",
+    srcs = ["cbits/impl.c"],
+)
+
+cc_library(
+    name = "cbits",
+    srcs = ["cbits/intf.c"],
+    deps = ["cbits-indirect"],
+)
+
+
+haskell_library(
+    name = "mypkg",
+    srcs = ["src/MyModule.hs"],
+    src_strip_prefix = "src",
+    deps = [":cbits", ":base"],
+)
+
+haskell_binary(
+    name = "mytest",
+    srcs = ["test/Main.hs"],
+    src_strip_prefix = "test",
+    deps = [":mypkg", ":base"],
+)

--- a/tests/indirect-link/cbits/impl.c
+++ b/tests/indirect-link/cbits/impl.c
@@ -1,0 +1,9 @@
+static int thing;
+
+int real_get_thing(void) {
+  return thing;
+}
+
+void real_set_thing(int value) {
+  thing = value;
+}

--- a/tests/indirect-link/cbits/intf.c
+++ b/tests/indirect-link/cbits/intf.c
@@ -1,0 +1,10 @@
+extern int real_get_thing(void);
+extern void real_set_thing(int value);
+
+int get_thing(void) {
+  return real_get_thing();
+}
+
+void set_thing(int value) {
+  real_set_thing(value);
+}

--- a/tests/indirect-link/src/MyModule.hs
+++ b/tests/indirect-link/src/MyModule.hs
@@ -1,0 +1,11 @@
+module MyModule where
+
+foreign import ccall get_thing :: IO Int
+
+getThing :: IO Int
+getThing = get_thing
+
+foreign import ccall set_thing :: Int -> IO ()
+
+setThing :: Int -> IO ()
+setThing = set_thing

--- a/tests/indirect-link/test/Main.hs
+++ b/tests/indirect-link/test/Main.hs
@@ -1,0 +1,9 @@
+module Main (main) where
+
+import qualified MyModule
+
+main :: IO ()
+main = do
+  print =<< MyModule.getThing
+  MyModule.setThing 123
+  print =<< MyModule.getThing

